### PR TITLE
SWARM-1828: initial implementation of the mission

### PR DESCRIPTION
**Pull request by michalszynkiewicz**

### CODE REVIEW

File: greeting-service/pom.xml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  <version>1.0-SNAPSHOT</version>
+  <packaging>war</packaging>
+
+  <name>WildFly Swarm - Cache Booster - Greeting Service</name>
+
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>org.wildfly.swarm</groupId>
+        <artifactId>bom-all</artifactId>
```
Comment (Ladicek) [inline greeting-service/pom.xml]: Please just `bom`. Makes it easier to keep the community and product boosters in sync.

File: greeting-service/pom.xml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
+  <dependencies>
+    <dependency>
+      <groupId>javax</groupId>
+      <artifactId>javaee-api</artifactId>
```
Comment (Ladicek) [inline greeting-service/pom.xml]: Why is this needed, if you depend on the fractions directly?

File: greeting-service/pom.xml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      <artifactId>jaxrs</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.wildfly.swarm</groupId>
+      <artifactId>cdi</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.infinispan</groupId>
+      <artifactId>infinispan-client-hotrod</artifactId>
+      <version>8.2.8.Final</version>
```
Comment (Ladicek) [inline greeting-service/pom.xml]: Does this correspond to the JDG server version we use? Also the version number should probably be in `<properties>` for easier updating in the future.

File: greeting-service/pom.xml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                  <include>wildfly-swarm</include>
+                </includes>
+                <excludes>
+                  <exclude>webapp</exclude>
+                </excludes>
+              </generator>
+              <enricher>
+                <config>
+                  <wildfly-swarm-health-check>
+                    <path>/api/alive</path>
```
Comment (Ladicek) [inline greeting-service/pom.xml]: It's enough to add the `monitor` fraction, the Fabric8 Maven plugin then takes care of the rest.

File: greeting-service/src/main/java/io/openshift/booster/cache/greeting/AliveEndpoint.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * Date: 2/12/18
+ */
+@Path("/alive")
+public class AliveEndpoint {
+
+    @GET
+    public Response isAlive() {
+        return Response.ok().build();
+    }
+}
```
Comment (Ladicek) [inline greeting-service/src/main/java/io/openshift/booster/cache/greeting/AliveEndpoint.java]: Can delete this when the `monitor` fraction is added.

File: greeting-service/src/main/java/io/openshift/booster/cache/greeting/NameCache.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    private static final Logger log = Logger.getLogger(NameCache.class);
+    private static final String KEY = "cute-name";
+
+    @Inject
+    @ConfigurationValue("infinispan.host")
+    private String infispanHost;
+    @Inject
+    @ConfigurationValue("infinispan.port")
+    private String infispanPort;
```
Comment (Ladicek) [inline greeting-service/src/main/java/io/openshift/booster/cache/greeting/NameCache.java]: Can make it an `int`?

File: greeting-service/src/main/java/io/openshift/booster/cache/greeting/NameService.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * @author Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ * <br>
+ * Date: 3/12/18
+ */
+@ApplicationScoped
+public class NameService {
+
+    @Inject
+    @ConfigurationValue("nameService.url")
```
Comment (Ladicek) [inline greeting-service/src/main/java/io/openshift/booster/cache/greeting/NameService.java]: We typically don't use camelCase for configurables.

File: greeting-service/src/main/resources/project-stages.yml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+nameService:
+  url: "http://wfswarm-cache-name:8080/api"
+infinispan:
+  host: cache-server
+  port: 11222
+---
```
Comment (Ladicek) [inline greeting-service/src/main/resources/project-stages.yml]: Please separate into individual `project-*.yml` files. Using a single `project-stages.yml` is deprecated.

File: name-service/pom.xml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  </parent>
+  <artifactId>wfswarm-cache-name</artifactId>
+  <packaging>war</packaging>
+
+  <name>WildFly Swarm - Cache Booster - Name Service</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>javax</groupId>
+      <artifactId>javaee-api</artifactId>
```
Comment (Ladicek) [inline name-service/pom.xml]: Again, I think this isn't needed if you depend on the individual fractions directly.

File: name-service/pom.xml [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                  <include>wildfly-swarm</include>
+                </includes>
+                <excludes>
+                  <exclude>webapp</exclude>
+                </excludes>
+              </generator>
+              <enricher>
+                <config>
+                  <wildfly-swarm-health-check>
+                    <path>/api/alive</path>
```
Comment (Ladicek) [inline name-service/pom.xml]: Same as above, add the `monitor` fraction and let FMP deal with health checks on its own.

File: name-service/src/main/java/io/openshift/booster/cache/name/AliveEndpoint.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+ * Date: 2/12/18
+ */
+@Path("/alive")
+public class AliveEndpoint {
+
+    @GET
+    public Response isAlive() {
+        return Response.ok().build();
+    }
+}
```
Comment (Ladicek) [inline name-service/src/main/java/io/openshift/booster/cache/name/AliveEndpoint.java]: Same as above, this can be deleted.

File: name-service/src/main/java/io/openshift/booster/cache/name/NameResource.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        log.info("generating the name");
+        performSlowOperation();
+        return new NameDto(UserNameGenerator.generate());
+    }
+
+    private void performSlowOperation() {
+        try {
+            Thread.sleep(2000L);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```
Comment (Ladicek) [inline name-service/src/main/java/io/openshift/booster/cache/name/NameResource.java]: Just... don't.

File: name-service/src/main/resources/project-stages.yml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+---
```
Comment (Ladicek) [inline name-service/src/main/resources/project-stages.yml]: Same as above, separate to individual `project-*.yml` files. Or in this case, just remove the `project: \n stage: local` stuff and rename the file to `project-local.yml`.

File: pom.xml [hunk 14]
```diff
@@ -0,0 +1,84 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>io.openshift</groupId>
+    <artifactId>booster-parent</artifactId>
+    <version>10</version>
```
Comment (Ladicek) [inline pom.xml]: Please move to `20`.

File: pom.xml [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+  <groupId>io.openshift.booster</groupId>
+  <artifactId>wfswarm-cache-booster</artifactId>
+  <version>1.0-SNAPSHOT</version>
+  <packaging>pom</packaging>
+  <properties>
+    <version.wildfly.swarm>2018.3.3</version.wildfly.swarm>
+    <maven.compiler.source>1.8</maven.compiler.source>
+    <maven.compiler.target>1.8</maven.compiler.target>
+    <version.arquillian.cube>1.9.1</version.arquillian.cube>
```
Comment (Ladicek) [inline pom.xml]: Please remove this and the `arquillian-cube-bom` from `<dependencyManagement>`, this is already brought in by latest `booster-parent`.

File: tests/src/test/java/io/openshift/booster/OpenshiftIT.java [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+    @RouteURL(NAME_SERVICE_APP)
+    private URL nameServiceUrl;
+
+    @RouteURL(GREETING_SERVICE_APP)
+    private URL greetingServiceUrl;
+
+    @Before
+    public void setup() {
+        await().pollInterval(1, TimeUnit.SECONDS).atMost(5, TimeUnit.MINUTES).until(() -> {
```
Comment (Ladicek) [inline tests/src/test/java/io/openshift/booster/OpenshiftIT.java]: Remove the entire `await` here and add `@AwaitRoute` to the `@RouteURL` fields. See here: https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-circuit-breaker-redhat/pull/3/files#diff-2fb1675d7aaf8d8a12dad4ffc0660157

File: tests/src/test/java/io/openshift/booster/OpenshiftIT.java [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        getGreeting();
+        long time = System.currentTimeMillis() - start;
+        assertThat("Server didn't respond fast enough. Expecting response in 1000 ms, got in " + time,
+                time,
+                lessThanOrEqualTo(1000L));
+    }
+
+    @Test
+    public void shouldClearCache() {
+        when().get(greetingServiceUrl + "api/cached")
```
Comment (Ladicek) [inline tests/src/test/java/io/openshift/booster/OpenshiftIT.java]: I can't really force you, but I like this formatting:

```java
when()
        .get(greetingServiceUrl + "api/cached")
.then()
        .body("cached", is(false));
```

etc. :-)

File: greeting-service/.openshift/cacheserver.yaml [hunk 18]
```diff
@@ -0,0 +1,79 @@
+# copied from https://github.com/snowdrop/spring-boot-cache-booster/blob/master/greeting-service/.openshiftio/cacheserver.yml
```
Comment (Ladicek) [inline greeting-service/.openshift/cacheserver.yaml]: The file needs to be in a directory called `.openshiftio` (note the _io_ at the end), and needs to be called `service.*.yaml`. I'd personally call it `service.jdg.yaml`, but `service.cacheserver.yaml` works too.

File: greeting-service/pom.xml (hunk 1 above)
Comment (Ladicek) [reply greeting-service/pom.xml]: Also depMan for `bom` should be inherited from the top-level POM, right? So the `<dependencyManagement>` section here is just useless, right?

**Review by Ladicek — CHANGES_REQUESTED** [review]

File: greeting-service/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply greeting-service/pom.xml]: done

File: greeting-service/pom.xml (hunk 2 above)
Comment (michalszynkiewicz) [reply greeting-service/pom.xml]: done

File: greeting-service/pom.xml (hunk 3 above)
Comment (michalszynkiewicz) [reply greeting-service/pom.xml]: If I remember correctly, this is the version that our Infinispan fraction is using, I'll try to align it with the JDG server.

File: pom.xml (hunk 14 above)
Comment (michalszynkiewicz) [reply pom.xml]: done

File: tests/src/test/java/io/openshift/booster/OpenshiftIT.java (hunk 16 above)
Comment (michalszynkiewicz) [reply tests/src/test/java/io/openshift/booster/OpenshiftIT.java]: done

File: greeting-service/src/main/java/io/openshift/booster/cache/greeting/AliveEndpoint.java (hunk 5 above)
Comment (michalszynkiewicz) [reply greeting-service/src/main/java/io/openshift/booster/cache/greeting/AliveEndpoint.java]: done

File: name-service/src/main/java/io/openshift/booster/cache/name/NameResource.java (hunk 12 above)
Comment (michalszynkiewicz) [reply name-service/src/main/java/io/openshift/booster/cache/name/NameResource.java]: done

File: name-service/src/main/resources/project-stages.yml (hunk 13 above)
Comment (michalszynkiewicz) [reply name-service/src/main/resources/project-stages.yml]: done

File: greeting-service/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply greeting-service/pom.xml]: done

File: greeting-service/pom.xml (hunk 2 above)
Comment (michalszynkiewicz) [reply greeting-service/pom.xml]: removed

File: name-service/pom.xml (hunk 10 above)
Comment (michalszynkiewicz) [reply name-service/pom.xml]: done

File: pom.xml (hunk 15 above)
Comment (michalszynkiewicz) [reply pom.xml]: done

File: greeting-service/src/main/java/io/openshift/booster/cache/greeting/NameService.java (hunk 7 above)
Comment (michalszynkiewicz) [reply greeting-service/src/main/java/io/openshift/booster/cache/greeting/NameService.java]: Would name-service be better?

File: greeting-service/src/main/resources/project-stages.yml (hunk 8 above)
Comment (michalszynkiewicz) [reply greeting-service/src/main/resources/project-stages.yml]: done

File: name-service/pom.xml (hunk 9 above)
Comment (michalszynkiewicz) [reply name-service/pom.xml]: removed

File: greeting-service/pom.xml (hunk 4 above)
Comment (michalszynkiewicz) [reply greeting-service/pom.xml]: done

File: tests/src/test/java/io/openshift/booster/OpenshiftIT.java (hunk 17 above)
Comment (michalszynkiewicz) [reply tests/src/test/java/io/openshift/booster/OpenshiftIT.java]: done

File: greeting-service/src/main/java/io/openshift/booster/cache/greeting/NameCache.java (hunk 6 above)
Comment (michalszynkiewicz) [reply greeting-service/src/main/java/io/openshift/booster/cache/greeting/NameCache.java]: it works :O

File: greeting-service/pom.xml (hunk 3 above)
Comment (michalszynkiewicz) [reply greeting-service/pom.xml]: I checked JDG 7.1.0 (the one in the image) and it uses Infinispan 8.4.0... but I cannot find such community version of Hotrod.
I tried the newest one, 9.2.0.Final, but it was failing.
There's also 8.2.10, I think it's the newest from 8.* branch.

File: greeting-service/.openshift/cacheserver.yaml (hunk 18 above)
Comment (michalszynkiewicz) [reply greeting-service/.openshift/cacheserver.yaml]: done

File: greeting-service/.openshiftio/application.yaml [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+    tags: wildfly-swarm, java, microservice
+    template.openshift.io/provider-display-name: "Red Hat, Inc."
+    template.openshift.io/documentation-url: "http://appdev.prod-preview.openshift.io/docs/wf-swarm-runtime.html#mission-cache-wf-swarm"
+    description: >-
+      The Cache Service Mission demonstrates caching with Infinispan.
+parameters:
+- name: SOURCE_REPOSITORY_URL
+  description: The source URL for the application
+  displayName: Source URL
+  value: https://github.com/michalszynkiewicz/wfswarm-cache.git
```
Comment (mkouba) [inline greeting-service/.openshiftio/application.yaml]: Shouldn't it be https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-cache.git?

Comment (michalszynkiewicz) [reply greeting-service/.openshiftio/application.yaml]: It sure should, fixed

**Review by kenfinnigan — APPROVED** [review]
